README.md
# TokFM Parser [![Build Status](https://travis-ci.org/slq/tokfm-parser.svg?branch=master)](https://travis-ci.org/slq/tokfm-parser) ![Travis](https://img.shields.io/badge/license-GPL-ff69b4.svg)
Parser / downloader for tokfm audition archives.

### Usage

```java
git clone https://github.com/slq/tokfm-parser
cd tokfm-parser
./gradlew fatJar
java -jar build/libs/tokfm-all-1.0-SNAPSHOT.jar
```

Download will start immediately with the following console output:

```
~/dev/java/tokfm-parser (master)
λ java -jar build\libs\tokfm-all-1.0-SNAPSHOT.jar
2017-06-14 0620 - Pierwsze Śniadanie w TOK-u - Portoryko stanie się 51. stanem USA. Mówi Lech Miodek.mp3                                               : 5,19
2017-06-14 0600 - Pierwsze Śniadanie w TOK-u - WWF Rżnięcie puszczy - bezzasadne. O wynikach raportu mówi Dariusz Gatkowski.mp3                        : 7,04
2017-06-14 0500 - Pierwsze Śniadanie w TOK-u - Poranny przegląd prasy.mp3                                                                              : 16,58
2017-06-13 2300 - Wieczór - Tomasz Stawiszyński - Czy wolno jeść franciszkanów, czy wolno jeść zwierzęta. Spór między kreacjonizmem i ewolucjonizmem.m : 23,98
2017-06-13 2200 - Godzina Filozofów - O wynajmowaniu, a nie posiadaniu - jak zmienia się kapitalizm.mp3                                                : 22,27
2017-06-13 2120 - Wieczór - Tomasz Stawiszyński - Napięcie polityczne wokół Korei.mp3                                                                  : 13,76
2017-06-13 2000 - Salon - Goście Passenta (Daniel Passent) - Goście 'Salonu' Anna Materska-Sosnowska, Jacek Kucharczyk, Jarosław Gugała.mp3            : 23,57
2017-06-13 1940 - Codzienny Magazyn Motoryzacyjny - Test Ford Kuga po faceliftingu.mp3                                                                 : 5,91
2017-06-13 1920 - Jeszcze Więcej Sportu - Rozgrywki tenisowe w nadchodzącym sezonie.mp3                                                                : 7,71
2017-06-13 1900 - Post Factum - Paweł Sulik - Po wyborach Theresa May negocjuje z opozycją i spotyka się z Emmanuelem Macron.mp3                       : 4,58
2017-06-13 1840 - Post Factum - Paweł Sulik - Historia uwięzionego w Rosji Murada Amrieva.mp3                                                          : 7,27
2017-06-13 1820 - Post Factum - Paweł Sulik - Raport NIK pokazuje fatalny stan polskiego górnictwa.mp3                                                 : 5,77
2017-06-13 1800 - Post Factum - Paweł Sulik - Greenpeace blokuje wycinkę drzew w Puszczy Białowieskiej. Relacja naszego reportera Jakuba Medka..mp3    : 4,75
2017-06-13 1800 - Przewodnik Technologiczny TOK FM - Unikajcie logowania się na swoje konta na publicznych komputerach.mp3                             : 0,39
2017-06-13 1740 - Kultura Osobista - Filmowa uczta we Wrocławiu - co na Nowych Horyzontach.mp3                                                         : 6,71
2017-06-13 1700 - Godziny Szczytu - magazyn reporterów - Baśka miała fajne... wodowanie.mp3                                                            : 10,57
2017-06-13 1640 - Co się stało - Komisja Europejska wszczyna procedurę przeciwko Polsce.mp3                                                            : 7,14
2017-06-13 1600 - Popołudnie Radia TOK FM - Kulisy Kremla - o książce Michaiła Zygara opowiadała Anna Łabuszewska.mp3                                  : 10,46
2017-06-13 1540 - Popołudnie Radia TOK FM - Lokatorzy jeszcze poczekają, aż minie spektakl polityczny - o działaniach Komisji Weryfikacyjnej.mp3       : 6,47
2017-06-13 1520 - Popołudnie Radia TOK FM - Porażka populistycznego Ruchu Pięciu Gwiazd w wyborach samorządowych we Włoszech.mp3                       : 5,48...
```

Make sure your console is wide enough (at least 150 characters), so that download status is updated smoothly.

`TokFM-Parser` will update downloading status showing the amout of MB downloaded.

When file already exists, parser will exit. This default behaviour can be overriden by using `-s` or `--skip` flag, which will allow parser to continue downloading even if file already exists (existing file will be skipped, not overwritten).

###### Available options:
```java
-m Mask to use, for example: Ziobro|Zandberg|Sikorski
-s Skip existing file
```

### Filename matching
When `mask` is provided (using `-m` or `--mask` flag), it will be used for `REGEX` matching as `matchPattern` shown below:
```java
Pattern.compile(matchPattern, Pattern.CASE_INSENSITIVE)
```

For example:
```java
-m "Ziobro|Zandberg|Sikorski"
```
will match every podcasrt with either `Ziobro`, `Zandberg` or `Sikorski` mentioned in title.
```java
λ java -jar build\libs\tokfm-all-1.0-SNAPSHOT.jar -m "Ziobro|Zandberg|Sikorski"
Found matched text: (Zandberg)
2017-06-06 0720 - Poranek - Jan Wróbel - Powinniśmy handel niedzielnym ograniczać - Adrian Zandberg z Partii Razem.mp3                                 : 5,49
Found matched text: (Ziobro)
2017-05-24 1520 - Światopodgląd - Błaszczak i Ziobro przed Sejmem w sprawie Igora Stachowiaka.mp3                                                      : 5,43
Found matched text: (Ziobro)
2017-05-21 0900 - Świat się chwieje - Czy Ziobro jest karą za grzechy sędziów.mp3                                                                      : 24,15
Found matched text: (Zandberg)
2017-05-12 1200 - A teraz na poważnie - Zandberg Stanowisko PO ws. przyjmowania uchodźców jest nieprzyzwoite. Grzegorz Schetyna jest cynikiem..mp3     : 10,27
```



docs/release-notes.md
**0.0.2 (2017-06-15)** - [2 commits](https://github.com/slq/tokfm-parser/compare/v0.0.1...v0.0.2) by [Slawek Plamowski](http://github.com/slq) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.0.2-green.svg)](https://bintray.com/slq/tokfm-parser/basic/0.0.2)
 - No pull requests referenced in commit messages.

**0.0.2 (2017-06-15)** - no code changes (no commits) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.0.2-green.svg)](https://bintray.com/slq/tokfm-parser/basic/0.0.2)

**0.0.2 (2017-06-15)** - no code changes (no commits) - published to [![Bintray](https://img.shields.io/badge/Bintray-0.0.2-green.svg)](https://bintray.com/shipkit/examples/basic/0.0.2)



src/main/resources/tokfm.user.js
// ==UserScript==
// @include       http://audycje.tokfm.pl/*
// @description   Wyciąga bezpośrednie linki do plików mp3
// @author        Marek Drwota
// ==/UserScript==

function sub4(s ,j) {
  return s.substr(j*4,4);
}

Number.prototype.pad = function(size) {
      var s = String(this);
      if(typeof(size) !== "number"){size = 2;}

      while (s.length < size) {s = "0" + s;}
      return s;
    }

var MD5 = function (string) {
    function RotateLeft(lValue, iShiftBits) {
        return (lValue<<iShiftBits) | (lValue>>>(32-iShiftBits));
    }
    function AddUnsigned(lX,lY) {
        var lX4,lY4,lX8,lY8,lResult;
        lX8 = (lX & 0x80000000);
        lY8 = (lY & 0x80000000);
        lX4 = (lX & 0x40000000);
        lY4 = (lY & 0x40000000);
        lResult = (lX & 0x3FFFFFFF)+(lY & 0x3FFFFFFF);
        if (lX4 & lY4) {
            return (lResult ^ 0x80000000 ^ lX8 ^ lY8);
        }
        if (lX4 | lY4) {
            if (lResult & 0x40000000) {
                return (lResult ^ 0xC0000000 ^ lX8 ^ lY8);
            } else {
                return (lResult ^ 0x40000000 ^ lX8 ^ lY8);
            }
        } else {
            return (lResult ^ lX8 ^ lY8);
        }
    }
    function F(x,y,z) { return (x & y) | ((~x) & z); }
    function G(x,y,z) { return (x & z) | (y & (~z)); }
    function H(x,y,z) { return (x ^ y ^ z); }
    function I(x,y,z) { return (y ^ (x | (~z))); }
    function FF(a,b,c,d,x,s,ac) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(F(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    };
    function GG(a,b,c,d,x,s,ac) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(G(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    };
    function HH(a,b,c,d,x,s,ac) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(H(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    };
    function II(a,b,c,d,x,s,ac) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(I(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    };
    function ConvertToWordArray(string) {
        var lWordCount;
        var lMessageLength = string.length;
        var lNumberOfWords_temp1=lMessageLength + 8;
        var lNumberOfWords_temp2=(lNumberOfWords_temp1-(lNumberOfWords_temp1 % 64))/64;
        var lNumberOfWords = (lNumberOfWords_temp2+1)*16;
        var lWordArray=Array(lNumberOfWords-1);
        var lBytePosition = 0;
        var lByteCount = 0;
        while ( lByteCount < lMessageLength ) {
            lWordCount = (lByteCount-(lByteCount % 4))/4;
            lBytePosition = (lByteCount % 4)*8;
            lWordArray[lWordCount] = (lWordArray[lWordCount] | (string.charCodeAt(lByteCount)<<lBytePosition));
            lByteCount++;
        }
        lWordCount = (lByteCount-(lByteCount % 4))/4;
        lBytePosition = (lByteCount % 4)*8;
        lWordArray[lWordCount] = lWordArray[lWordCount] | (0x80<<lBytePosition);
        lWordArray[lNumberOfWords-2] = lMessageLength<<3;
        lWordArray[lNumberOfWords-1] = lMessageLength>>>29;
        return lWordArray;
    };
    function WordToHex(lValue) {
        var WordToHexValue="",WordToHexValue_temp="",lByte,lCount;
        for (lCount = 0;lCount<=3;lCount++) {
            lByte = (lValue>>>(lCount*8)) & 255;
            WordToHexValue_temp = "0" + lByte.toString(16);
            WordToHexValue = WordToHexValue + WordToHexValue_temp.substr(WordToHexValue_temp.length-2,2);
        }
        return WordToHexValue;
    };
    function Utf8Encode(string) {
        string = string.replace(/\r\n/g,"\n");
        var utftext = "";
        for (var n = 0; n < string.length; n++) {
            var c = string.charCodeAt(n);
            if (c < 128) {
                utftext += String.fromCharCode(c);
            }
            else if((c > 127) && (c < 2048)) {
                utftext += String.fromCharCode((c >> 6) | 192);
                utftext += String.fromCharCode((c & 63) | 128);
            }
            else {
                utftext += String.fromCharCode((c >> 12) | 224);
                utftext += String.fromCharCode(((c >> 6) & 63) | 128);
                utftext += String.fromCharCode((c & 63) | 128);
            }
        }
        return utftext;
    };
    var x=Array();
    var k,AA,BB,CC,DD,a,b,c,d;
    var S11=7, S12=12, S13=17, S14=22;
    var S21=5, S22=9 , S23=14, S24=20;
    var S31=4, S32=11, S33=16, S34=23;
    var S41=6, S42=10, S43=15, S44=21;
    string = Utf8Encode(string);
    x = ConvertToWordArray(string);
    a = 0x67452301; b = 0xEFCDAB89; c = 0x98BADCFE; d = 0x10325476;
    for (k=0;k<x.length;k+=16) {
        AA=a; BB=b; CC=c; DD=d;
        a=FF(a,b,c,d,x[k+0], S11,0xD76AA478);
        d=FF(d,a,b,c,x[k+1], S12,0xE8C7B756);
        c=FF(c,d,a,b,x[k+2], S13,0x242070DB);
        b=FF(b,c,d,a,x[k+3], S14,0xC1BDCEEE);
        a=FF(a,b,c,d,x[k+4], S11,0xF57C0FAF);
        d=FF(d,a,b,c,x[k+5], S12,0x4787C62A);
        c=FF(c,d,a,b,x[k+6], S13,0xA8304613);
        b=FF(b,c,d,a,x[k+7], S14,0xFD469501);
        a=FF(a,b,c,d,x[k+8], S11,0x698098D8);
        d=FF(d,a,b,c,x[k+9], S12,0x8B44F7AF);
        c=FF(c,d,a,b,x[k+10],S13,0xFFFF5BB1);
        b=FF(b,c,d,a,x[k+11],S14,0x895CD7BE);
        a=FF(a,b,c,d,x[k+12],S11,0x6B901122);
        d=FF(d,a,b,c,x[k+13],S12,0xFD987193);
        c=FF(c,d,a,b,x[k+14],S13,0xA679438E);
        b=FF(b,c,d,a,x[k+15],S14,0x49B40821);
        a=GG(a,b,c,d,x[k+1], S21,0xF61E2562);
        d=GG(d,a,b,c,x[k+6], S22,0xC040B340);
        c=GG(c,d,a,b,x[k+11],S23,0x265E5A51);
        b=GG(b,c,d,a,x[k+0], S24,0xE9B6C7AA);
        a=GG(a,b,c,d,x[k+5], S21,0xD62F105D);
        d=GG(d,a,b,c,x[k+10],S22,0x2441453);
        c=GG(c,d,a,b,x[k+15],S23,0xD8A1E681);
        b=GG(b,c,d,a,x[k+4], S24,0xE7D3FBC8);
        a=GG(a,b,c,d,x[k+9], S21,0x21E1CDE6);
        d=GG(d,a,b,c,x[k+14],S22,0xC33707D6);
        c=GG(c,d,a,b,x[k+3], S23,0xF4D50D87);
        b=GG(b,c,d,a,x[k+8], S24,0x455A14ED);
        a=GG(a,b,c,d,x[k+13],S21,0xA9E3E905);
        d=GG(d,a,b,c,x[k+2], S22,0xFCEFA3F8);
        c=GG(c,d,a,b,x[k+7], S23,0x676F02D9);
        b=GG(b,c,d,a,x[k+12],S24,0x8D2A4C8A);
        a=HH(a,b,c,d,x[k+5], S31,0xFFFA3942);
        d=HH(d,a,b,c,x[k+8], S32,0x8771F681);
        c=HH(c,d,a,b,x[k+11],S33,0x6D9D6122);
        b=HH(b,c,d,a,x[k+14],S34,0xFDE5380C);
        a=HH(a,b,c,d,x[k+1], S31,0xA4BEEA44);
        d=HH(d,a,b,c,x[k+4], S32,0x4BDECFA9);
        c=HH(c,d,a,b,x[k+7], S33,0xF6BB4B60);
        b=HH(b,c,d,a,x[k+10],S34,0xBEBFBC70);
        a=HH(a,b,c,d,x[k+13],S31,0x289B7EC6);
        d=HH(d,a,b,c,x[k+0], S32,0xEAA127FA);
        c=HH(c,d,a,b,x[k+3], S33,0xD4EF3085);
        b=HH(b,c,d,a,x[k+6], S34,0x4881D05);
        a=HH(a,b,c,d,x[k+9], S31,0xD9D4D039);
        d=HH(d,a,b,c,x[k+12],S32,0xE6DB99E5);
        c=HH(c,d,a,b,x[k+15],S33,0x1FA27CF8);
        b=HH(b,c,d,a,x[k+2], S34,0xC4AC5665);
        a=II(a,b,c,d,x[k+0], S41,0xF4292244);
        d=II(d,a,b,c,x[k+7], S42,0x432AFF97);
        c=II(c,d,a,b,x[k+14],S43,0xAB9423A7);
        b=II(b,c,d,a,x[k+5], S44,0xFC93A039);
        a=II(a,b,c,d,x[k+12],S41,0x655B59C3);
        d=II(d,a,b,c,x[k+3], S42,0x8F0CCC92);
        c=II(c,d,a,b,x[k+10],S43,0xFFEFF47D);
        b=II(b,c,d,a,x[k+1], S44,0x85845DD1);
        a=II(a,b,c,d,x[k+8], S41,0x6FA87E4F);
        d=II(d,a,b,c,x[k+15],S42,0xFE2CE6E0);
        c=II(c,d,a,b,x[k+6], S43,0xA3014314);
        b=II(b,c,d,a,x[k+13],S44,0x4E0811A1);
        a=II(a,b,c,d,x[k+4], S41,0xF7537E82);
        d=II(d,a,b,c,x[k+11],S42,0xBD3AF235);
        c=II(c,d,a,b,x[k+2], S43,0x2AD7D2BB);
        b=II(b,c,d,a,x[k+9], S44,0xEB86D391);
        a=AddUnsigned(a,AA);
        b=AddUnsigned(b,BB);
        c=AddUnsigned(c,CC);
        d=AddUnsigned(d,DD);
    }
    var temp = WordToHex(a)+WordToHex(b)+WordToHex(c)+WordToHex(d);
    return temp.toUpperCase();
}



function zm() {
var linki=document.getElementsByTagName('li');
for (i=0;i<linki.length;i+=1)
  if (linki[i].hasAttribute('class'))
    if (linki[i].getAttribute('class') == 'entry')
      if (linki[i].innerHTML.indexOf('data-id="') > 0)
      {
	file_id=linki[i].innerHTML.substr(linki[i].innerHTML.indexOf('data-id="')+9,10);
	file_id=file_id.substr(0,file_id.indexOf('"'));
	tmpid=parseInt(file_id);
	a=tmpid%100;
	tmpid = Math.floor(tmpid/100);
	b=tmpid%100;
	tmpid = Math.floor(tmpid/100);
	c=tmpid%100;
	tmpid = Math.floor(tmpid/100);
	d=tmpid%100;
	e = Math.floor(tmpid/100);
	file_name=e.pad()+'/'+d.pad()+'/'+c.pad()+'/'+b.pad()+'/'+a.pad()+'.mp3';
	d = new Date();
	var n = d.getTime();
	n=Math.floor(n/1000);
	n=n.toString(16).toUpperCase();
	mdp=MD5("MwbJdy3jUC2xChua/" + file_name + n);
	load=n + "." + sub4(mdp,3) + "." + sub4(mdp,2) + "." + sub4(mdp,4) + "." + sub4(mdp,5) + ".";
	token=Math.round(Math.random() * 255).toString(16).toUpperCase() + "." + sub4(mdp,0) + "." + sub4(mdp,1) + "." + sub4(mdp,7) + "." + sub4(mdp,6) + ".";
	linki[i].innerHTML = linki[i].innerHTML + 'wget --post-data="pre=load" --header="X-Tuba:' + file_name + '" --header="X-Tuba-Load:' + load + '" --header="X-Tuba-Token:' + token +'" http://storage.tuba.fm/load_podcast/' + file_id + '.mp3';
	linki[i].setAttribute('class', 'downloadable');

      }

window.setTimeout(zm,5000);
}
window.setTimeout(zm,1000);